### PR TITLE
Emv first part of commands

### PR DESCRIPTION
**Pull request by merlokk**
EMV commands
proxmark3> hf emv
help             This help
exec             Executes EMV contactless transaction.
pse              Execute PPSE. It selects 2PAY.SYS.DDF01 or 1PAY.SYS.DDF01 directory.
search           Try to select all applets from applets list and print installed applets.
select           Select applet.

command exec works up to select application stage.

### CODE REVIEW

**Comment by merlokk** [comment]
try 'hf emv exec -s -t -a' ))
done.

**Comment by iceman1001** [comment]
I'm impressed with the amount of functionality you add.

Pretty soon the discussion of a unified parameter is needed,   since you seem to be fond of the prefix "-" 
while we have spend time not using it.  I don't mind either,  but I prefer that we have one and not three like we have to day.  
1.  no prefixed inparameters.
2.  letter only prefixed parameters
3.  "-" + letter prefixed parameters.

Same goes for inputed "hex"..  Either  "405060"  or "40 50 60"  or both accepted.

**Comment by iceman1001** [comment]
Seems ok to merge.

**Comment by pwpiwi** [comment]
> Pretty soon the discussion of a unified parameter is needed

Ah,  yes, please. I will open a separate issue...

**Comment by iceman1001** [comment]
So merge?

**Comment by pwpiwi** [comment]
Didn't check in depth yet.

**Comment by merlokk** [comment]
Check wo merge, i will merge soon. too much cross links...  merge part into tlv printing, so the rest functions can be tested)

**Comment by iceman1001** [comment]
@merlokk  Will you update this PR?

**Comment by merlokk** [comment]
Yes, I need my PC with 2 monitors) I will update it at monday.
Now this PR can be tested as is (if it needs) because it have tlv decoder similar with master's

**Comment by merlokk** [comment]
not ready. needs to fix look

**Comment by merlokk** [comment]
ready to merge

**Comment by merlokk** [comment]
merge?

**Comment by merlokk** [comment]
lets merge) i have next part)

**Comment by merlokk** [comment]
Lets merge

**Comment by pwpiwi** [comment]
A few questions: How much emv is in the emv commands? Isn't most of it ISO7816? And isn't it independent of the lower layer? I.e. shouldn't it also work with ISO14443B cards?

What I am aiming at:
1. there are only a few EMV specific commands. Selecting the predefined 2PAY.SYS.DDF01 or 1PAY.SYS.DDF01 for example. The generic commands however should be called e.g. `hf 7816 ...`
2. one parameter of the commands should select the underlying protocol, i.e. 14a or 14b (15 as well?) instead of having it fixed to 14a.

What do you think?

**Comment by merlokk** [comment]
There are will be many emv commands. And Yes, dual interface cards can answer via 7816 and 14443. general part - APDU. Emv laid on APDU. 
And cards May answer differently via 7816 and 14443

And it should work via 14443a and B, but now I work on A because it wide used. 
I will add B later.

**Comment by merlokk** [comment]
Like that
7816 (T=0, T=1), 14443 (T=CL) - > APDU - > EMV

**Comment by pwpiwi** [comment]
I don't mean the physical interfaces. I mean ISO7816-4 which specifies BER-TLV.

**Comment by merlokk** [comment]
there are an ISO commands
http://www.cardwerk.com/smartcards/smartcard_standard_ISO7816-4_6_basic_interindustry_commands.aspx
and EMV based on them, but not fully. As sample:
```
Select:
EMV - 00 A4 04 00 data
ISO - XX A4 0X 0X
GetChallenge:
EMV - 00 84 00 00 00
ISO - XX B4 00 00
```
EMV is based on ISO commands and there are many ISO commands that not fully supported by EMV cards and EMV cards supported only a subset of command's parameters.
We can do `hf iso` commands and make there full set of ISO commands. Now there is `hf 14a raw`. 
But EMV is a  subset (and sometimes not a standard...) of ISO commands.

**Comment by merlokk** [comment]
I found good cross-standard table with APDU commands 
http://techmeonline.com/most-used-smart-card-commands-apdu/

**Comment by merlokk** [comment]
lets merge. I have second part of emv and command line parser to discuss.
it will be easier for me (in git) to have this PR merged to master.
If you want to change something - lets change after.

File: client/emv/cmdemv.c [hunk 1]
```diff
@@ -0,0 +1,537 @@
...
+// the license.
+//-----------------------------------------------------------------------------
+// EMV commands
+//-----------------------------------------------------------------------------
+
+#include "cmdemv.h"
+
+int UsageCmdHFEMVSelect(void) {
+	PrintAndLog("HELP :  Executes select applet command:\n");
+	PrintAndLog("Usage:  hf emv select [-s][-k][-1][-2][-t] <HEX applet AID>\n");
```
Comment (pwpiwi) [inline client/emv/cmdemv.c]: Options -1 and -2 ?

File: client/emv/cmdemv.c [hunk 2]
```diff
@@ -0,0 +1,537 @@
...
+	}
+	
+	// exec
+	uint8_t buf[200] = {0};
+	size_t len = 0;
+	uint16_t sw = 0;
+	int res = EMVSelect(activateField, leaveSignalON, data, datalen, buf, sizeof(buf), &len, &sw, NULL);
+
+	if (sw)
+		PrintAndLog("APDU response: %04x - %s", sw, GetAPDUCodeDescription(sw >> 8, sw & 0xff));
```
Comment (pwpiwi) [inline client/emv/cmdemv.c]: Should say "Response Status"

File: client/emv/cmdemv.c [hunk 3]
```diff
@@ -0,0 +1,537 @@
...
+			// we get all the hex to end of line with spaces
+			break;
+		}
+
+
+		cmdp++;
+	}
+	
+	// exec
+	uint8_t buf[200] = {0};
```
Comment (pwpiwi) [inline client/emv/cmdemv.c]: Why 200? Is this sufficient in all cases?

File: client/emv/cmdemv.c [hunk 4]
```diff
@@ -0,0 +1,537 @@
...
+					break;
+				default:
+					PrintAndLog("Unknown parameter '%c'", param_getchar_indx(cmd, 1, cmdp));
+					return 1;
+		}
+		cmdp++;
+	}
+	
+	// exec
+	uint8_t buf[200] = {0};
```
Comment (pwpiwi) [inline client/emv/cmdemv.c]: See above.

File: client/emv/emvcore.c [hunk 5]
```diff
@@ -10,6 +10,77 @@
...
 
+// Got from here. Thanks)
+// https://eftlab.co.uk/index.php/site-map/knowledge-base/211-emv-aid-rid-pix
+const char *PSElist [] = { 
+	"325041592E5359532E4444463031", // 2PAY.SYS.DDF01 - Visa Proximity Payment System Environment - PPSE
+	"315041592E5359532E4444463031"  // 1PAY.SYS.DDF01 - Visa Payment System Environment - PSE
+};
+const size_t PSElistLen = sizeof(PSElist)/sizeof(char*);
+
+const char *AIDlist [] = {
```
Comment (pwpiwi) [inline client/emv/emvcore.c]: Duplicate of emv.h?

File: client/emv/emvcore.c [hunk 6]
```diff
@@ -23,11 +94,370 @@ void TLVPrintFromBuffer(uint8_t *data, int datalen) {
...
+	}
+	
+	// select
+	res = EMVSelect(ActivateField, LeaveFieldON, buf, len, Result, MaxResultLen, ResultLen, sw, NULL);
+
+	return res;
+}
+
+int EMVSearchPSE(bool ActivateField, bool LeaveFieldON, bool decodeTLV, struct tlvdb *tlv) {
+	uint8_t data[260] = {0};
```
Comment (pwpiwi) [inline client/emv/emvcore.c]: Why 260?

File: client/emv/emvcore.c [hunk 7]
```diff
@@ -23,11 +94,370 @@ void TLVPrintFromBuffer(uint8_t *data, int datalen) {
...
+	}
+	
+	if(!LeaveFieldON)
+		DropField();
+	
+	return res;
+}
+
+int EMVSearch(bool ActivateField, bool LeaveFieldON, bool decodeTLV, struct tlvdb *tlv) {
+	uint8_t aidbuf[50] = {0};
```
Comment (pwpiwi) [inline client/emv/emvcore.c]: ...

File: client/emv/emvcore.c (hunk 5 above)
Comment (merlokk) [reply client/emv/emvcore.c]: emv.h deleted. i forget it...

File: client/emv/cmdemv.c [hunk 8]
```diff
@@ -0,0 +1,537 @@
...
+	PrintAndLog("  Options:");
+	PrintAndLog("  -s       : select card");
+	PrintAndLog("  -k       : keep field for next command");
+	PrintAndLog("  -1       : ppse (1PAY.SYS.DDF01)");
+	PrintAndLog("  -2       : pse (2PAY.SYS.DDF01)");
+	PrintAndLog("  -t       : TLV decode results\n");
+	PrintAndLog("Samples:");
+	PrintAndLog(" hf emv pse -s 1 -> select, get pse");
+	PrintAndLog(" hf emv pse -s -k 2 -> select, get ppse, keep field");
+	PrintAndLog(" hf emv pse -s -t 2 -> select, get ppse, show result in TLV");
```
Comment (pwpiwi) [inline client/emv/cmdemv.c]: Examples are wrong. Must be -1 and -2 respectively.

**Comment by pwpiwi** [comment]
No results from hf emv search ?
```
proxmark3> hf emv search -s -t
Search completed.
proxmark3>
```

File: client/emv/cmdemv.c [hunk 9]
```diff
@@ -0,0 +1,537 @@
...
+
+	return 0;
+}
+
+int UsageCmdHFEMVExec(void) {
+	PrintAndLog("HELP :  Executes EMV contactless transaction:\n");
+	PrintAndLog("Usage:  hf emv exec [-s][-a][-t]\n");
+	PrintAndLog("  Options:");
+	PrintAndLog("  -s       : select card");
+	PrintAndLog("  -a       : show APDU reqests and responses\n");
```
Comment (pwpiwi) [inline client/emv/cmdemv.c]: Why do we have a -a option here and not in other emv commands?

**Comment by pwpiwi** [comment]
```
proxmark3> hf emv exec -s -t -a

* PPSE.
>>>> 00 a4 04 00 0e 32 50 41 59 2e 53 59 53 2e 44 44 46 30 31
<<<< 6f 36 84 0e 32 50 41 59 2e 53 59 53 2e 44 44 46 30 31 a5 24 bf 0c 21 61 1f 4f 08 a0 00 00 00 25 01 08 01 50 10 41 4d 45 52 49 43 41 4e 20 45 58 50 52 45 53 53 87 01 01 90 00
>>>> 00 a4 04 00 08 a0 00 00 00 25 01 08 01
<<<< 67 00
SELECT ERROR: [6700] Wrong length
a000000025010801:
TLV ERROR: Can't parse response as TLV tree.

* Search AID in list.
|------------------|--------|-------------------------|
|    AID           |Priority| Name                    |
|------------------|--------|-------------------------|
|                         none                        |
|------------------|--------|-------------------------|
Can't select AID. EMV AID not found
proxmark3>
```

File: client/emv/cmdemv.c (hunk 9 above)
Comment (merlokk) [reply client/emv/cmdemv.c]: Ill add it

Comment (merlokk) [reply client/emv/cmdemv.c]: or.......   the rest commands shows apdu by default. 
so add this option?

**Comment by merlokk** [comment]
may you issue hf emv pse -2 -t?

**Comment by merlokk** [comment]
and
`hf emv select -s -t A000000025010801`

File: client/emv/emvcore.c (hunk 6 above)
Comment (merlokk) [reply client/emv/emvcore.c]: fixed. 50 bytes = 100hex symbols. it enough for AID (ususally they have only 7-10 bytes)

File: client/emv/emvcore.c (hunk 7 above)
Comment (merlokk) [reply client/emv/emvcore.c]: I can unify this but it enoughfixed. 50 bytes = 100hex symbols. it enough for AID (ususally they have only 7-10 bytes)

File: client/emv/emvcore.c (hunk 6 above)
Comment (merlokk) [reply client/emv/emvcore.c]: fixed

**Comment by merlokk** [comment]
I have added Le==0 field try please
`hf emv select -s -t A000000025010801`

File: client/emv/cmdemv.c (hunk 9 above)
Comment (merlokk) [reply client/emv/cmdemv.c]: I can add this option only to search. search produces many lines to console...

File: client/emv/emvcore.c [hunk 10]
```diff
@@ -23,11 +94,370 @@ void TLVPrintFromBuffer(uint8_t *data, int datalen) {
...
+		
+		ttmp = tlvdb_find_next(ttmp, 0x6f);
+	}
+
+	PrintAndLog("|------------------|--------|-------------------------|");
+}
+
+
+int EMVSelect(bool ActivateField, bool LeaveFieldON, uint8_t *AID, size_t AIDLen, uint8_t *Result, size_t MaxResultLen, size_t *ResultLen, uint16_t *sw, struct tlvdb *tlv) {
+	uint8_t data[260] = {0};
```
Comment (pwpiwi) [inline client/emv/emvcore.c]: Here we have 260 again. Maybe there is a reason for it? E.g. 256 Bytes block size + 2 Response Status Bytes + 2 CRC Bytes added by iso14443a apdu?

File: client/emv/emvcore.c (hunk 6 above)
Comment (pwpiwi) [reply client/emv/emvcore.c]: No?

File: client/emv/cmdemv.c [hunk 11]
```diff
@@ -0,0 +1,540 @@
...
+			}
+			
+			// we get all the hex to end of line with spaces
+			break;
+		}
+
+
+		cmdp++;
+	}
+
```
Comment (pwpiwi) [inline client/emv/cmdemv.c]: Need to handle `datalen == 0` (no AID)

File: client/emv/cmdemv.c (hunk 9 above)
Comment (pwpiwi) [reply client/emv/cmdemv.c]: For the sake of consistency I propose to have it for all emv commands.

File: client/emv/emvcore.c (hunk 10 above)
Comment (merlokk) [reply client/emv/emvcore.c]: There was this reason, I have added more space to be sure.
There is one situation - chaining. In this situation APDU May be bigger. But now we dont support it on arm side and for emv it not needs. Cards with emv dont needs to support this sns ever simply not support.

File: client/emv/emvcore.c [hunk 12]
```diff
@@ -23,11 +94,370 @@ void TLVPrintFromBuffer(uint8_t *data, int datalen) {
...
+	data[4] = AIDLen;
+	memcpy(&data[5], AID, AIDLen);
+	
+	if (ActivateField)
+		DropField();
+	
+	if (APDULogging)
+		PrintAndLog(">>>> %s", sprint_hex(data, AIDLen + 6));
+
+	int res = ExchangeAPDU14a(data, AIDLen + 6, ActivateField, LeaveFieldON, Result, (int *)ResultLen);
```
Comment (pwpiwi) [inline client/emv/emvcore.c]: `MaxResultLen` is not forwarded to `ExchangeAPDU14a()`. In fact it is not used at all?

**Comment by pwpiwi** [comment]
`hf emv search` and `hf emv exec` work after adding the `Le` byte

**Comment by merlokk** [comment]
Collis. Im going to business trip to friday. Maybe can POST in 1-2 days

File: client/emv/cmdemv.c (hunk 11 above)
Comment (merlokk) [reply client/emv/cmdemv.c]: command select can select default application wo AID

**Comment by merlokk** [comment]
looks done) go to next flight from Dubai)

File: client/emv/cmdemv.c (hunk 11 above)
Comment (pwpiwi) [reply client/emv/cmdemv.c]: Indeed.

**Pull request merged**